recipes/openai_textcat_multilabel.py

```diff
@@ -0,0 +1,527 @@
+    nlp = spacy.blank(lang)
+
+    if segment:
+        nlp.add_pipe("sentencizer")
+
+    openai = OpenAISuggester(
+        openai_model=model,
+        labels=labels,
+        max_examples=max_examples,
+        prompt_template=_load_template(prompt_path),
+        exclusive_classes=exclusive_classes,
+        verbose=verbose,
+        segment=segment,
+        openai_api_key=api_key,
+        openai_api_org=api_org,
+        openai_timeout_s=120,
+        openai_n=10,
+    )
+
+    for eg in examples:
+        openai.add_example(eg)
+
+    stream = list(srsly.read_jsonl(input_path))
+    stream = openai(tqdm.tqdm(stream), batch_size=batch_size, nlp=nlp)
+
+    return {
+        "dataset": dataset,
+        "view_id": "blocks",
+        "stream": stream,
+        "update": openai.update,
+        "config": {
+            "labels": openai.labels,
+            "batch_size": batch_size,
+            "exclude_by": "input",
+            "choice_style": "single" if exclusive_classes else "multiple",
+            "blocks": [
+                {"view_id": "choice"},
+                {"view_id": "html", "html_template": HTML_TEMPLATE},
+            ],
+            "show_flag": True,
+            "global_css": CSS_FILE_PATH.read_text(),
+        },
+    }
+
+
+@prodigy.recipe(
+    "textcat_multilabel.openai.fetch",
+    input_path=("Path to jsonl data to annotate", "positional", None, Path),
+    output_path=("Path to save the output", "positional", None, Path),
+    labels=("Labels (comma delimited)", "positional", None, lambda s: s.split(",")),
+    lang=("Language to use for tokenizer.", "option", "l", str),
+    model=("GPT-3 model to use for completion", "option", "m", str),
+    examples_path=("Examples file to help define the task", "option", "e", Path),
+    max_examples=("Max examples to include in prompt", "option", "n", int),
+    prompt_path=("Path to jinja2 prompt template", "option", "p", Path),
+    batch_size=("Batch size to send to OpenAI API", "option", "b", int),
+    segment=("Split sentences", "flag", "S", bool),
+    verbose=("Print extra information to terminal", "option", "flag", bool),
+)
+def textcat_multilabel_openai_fetch(
+    input_path: Path,
+    output_path: Path,
+    labels: List[str],
+    lang: str = "en",
+    model: str = "text-davinci-003",
+    batch_size: int = 10,
+    segment: bool = False,
+    examples_path: Optional[Path] = None,
+    prompt_path: Path = DEFAULT_PROMPT_PATH,
+    max_examples: int = 2,
+    verbose: bool = False,
+):
+    """Get bulk TextCat suggestions from an OpenAI API, using zero-shot or few-shot learning.
+    The results can then be corrected using the `textcat.manual` recipe."""
+    api_key, api_org = _get_api_credentials(model)
+    examples = _read_prompt_examples(examples_path)
+    nlp = spacy.blank(lang)
+    if segment:
+        nlp.add_pipe("sentencizer")
+    openai = OpenAISuggester(
+        openai_model=model,
+        labels=labels,
```

Comment:
When I tried running: 

```
python -m prodigy textcat_multilabel.openai.fetch doubt.jsonl doubt-annot.jsonl "sarcasm,anger,joy"  -F recipes/openai_textcat_multilabel.py
```

I got this error. 

```
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/prodigy/__main__.py", line 61, in <module>
    controller = recipe(*args, use_plac=True)
  File "cython_src/prodigy/core.pyx", line 329, in prodigy.core.recipe.recipe_decorator.recipe_proxy
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/plac_core.py", line 367, in call
    cmd, result = parser.consume(arglist)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/plac_core.py", line 232, in consume
    return cmd, self.func(*(args + varargs + extraopts), **kwargs)
  File "recipes/openai_textcat_multilabel.py", line 386, in textcat_multilabel_openai_fetch
    openai = OpenAISuggester(
TypeError: __init__() missing 1 required keyword-only argument: 'exclusive_classes'
```

I've manually overridden it to true on my end but I think you'll want to add this as a param for the recipe.

Reply:
I also had a look at the output of this recipe and noticed something strange. 

Again, this is the recipe call: 

```
python -m prodigy textcat_multilabel.openai.fetch doubt.jsonl doubt-annot-ne.jsonl "sarcasm,anger,joy"  -F recipes/openai_textcat_multilabel.py
```

When I manually set `exclusive_classes=True` in `OpenAISuggester` then the output via `db-out` looks like this: 

```
> cat doubt-annot-excl.jsonl | jq -c "{accept: .accept}"
{"accept":["joy"]}
{"accept":["sarcasm"]}
{"accept":["joy"]}
{"accept":["sarcasm"]}
{"accept":["Sarcasm"]}
{"accept":["sarcasm"]}
{"accept":["Sarcasm"]}
{"accept":["joy"]}
{"accept":["sarcasm"]}
{"accept":["Sarcasm",""]}
...
```

When I manually set `exclusive_classes=False` in `OpenAISuggester` then the output via `db-out` looks like this: 

```
> cat doubt-annot-not-excl.jsonl | jq -c "{accept: .accept}"
{"accept":["joy","sarcasm"]}
{"accept":["sarcasm","joy",""]}
{"accept":["joy","sarcasm"]}
{"accept":["sarcasm","joy",""]}
{"accept":["sarcasm","joy"]}
{"accept":["sarcasm","joy"]}
{"accept":["sarcasm","joy"]}
{"accept":["joy","sarcasm"]}
{"accept":["sarcasm","joy",""]}
{"accept":["sarcasm","joy",""]}
...
```

It seems like it passes an empty string to the accept key here? That suggests something is going awry with parsing, but it also suggests (and Matt totally had to reminder me of this on my own recipe) that it might be good to write some unit tests that check this.

Reply:
Oh! And also notice how the labels can have inconsistent capitalisation. Might also be something to test for.

Reply:
Yeah, that those capitals are causing inconsistencies in the `textcat.manual` task that follows up.

```
python -m prodigy textcat.manual lj-review-1 doubt-annot.jsonl --label "sarcasm,anger,joy"  -F recipes/openai_textcat_multilabel.py
```

This leads to examples that are empty, like this one: 

![CleanShot 2023-01-17 at 13 24 40](https://user-images.githubusercontent.com/1019791/212898494-d6cb5b2b-56ee-436a-88f4-fdf9a9a429a0.png)